torchrl/data/tensor_specs.py

```diff
@@ -1014,61 +1026,80 @@ class MultiDiscreteTensorSpec(DiscreteTensorSpec):
 
     def __init__(
         self,
-        nvec: Sequence[int],
+        nvec: Union[Sequence[int], torch.Tensor, int],
         device: Optional[DEVICE_TYPING] = None,
         dtype: Optional[Union[str, torch.dtype]] = torch.long,
     ):
+        if not isinstance(nvec, torch.Tensor):
+            nvec = torch.tensor(nvec)
+        if nvec.ndim < 1:
+            nvec = nvec.unsqueeze(0)
+        self.nvec = nvec
         dtype, device = _default_dtype_and_device(dtype, device)
-        self._size = len(nvec)
-        shape = torch.Size([self._size])
-        space = BoxList([DiscreteBox(n) for n in nvec])
+        shape = nvec.shape
+        space = BoxList.from_nvec(nvec)
         super(DiscreteTensorSpec, self).__init__(
             shape, space, device, dtype, domain="discrete"
         )
 
+    def _rand(self, space: Box, shape: torch.Size):
+        x = []
+        for _s in space:
+            if isinstance(_s, BoxList):
+                x.append(self._rand(_s, shape))
+            else:
+                x.append(
+                    torch.randint(
+                        0,
+                        _s.n,
+                        shape,
+                        device=self.device,
+                        dtype=self.dtype,
+                    )
+                )
+        return torch.stack(x, -1)
+
     def rand(self, shape: Optional[torch.Size] = None) -> torch.Tensor:
         if shape is None:
             shape = torch.Size([])
-        x = torch.cat(
-            [
-                torch.randint(
-                    0,
-                    space.n,
-                    torch.Size([1, *shape]),
-                    device=self.device,
-                    dtype=self.dtype,
-                )
-                for space in self.space
-            ]
-        ).squeeze()
-        _size = [self._size] if self._size > 1 else []
-        return x.T.reshape([*shape, *_size])
 
-    def _split(self, val: torch.Tensor):
-        return [val] if self._size < 2 else val.split(1, -1)
+        x = self._rand(self.space, shape)
+        if self.nvec.ndim > 1:
+            x = x.transpose(len(shape), -1)
+        return x.squeeze(-1)
 
     def _project(self, val: torch.Tensor) -> torch.Tensor:
-        if val.dtype not in (torch.int, torch.long):
+        val_is_scalar = val.ndim < 1
+        if val_is_scalar:
+            val = val.unsqueeze(0)
```

Comment:
why do we unsqueeze and squeeze back? Can't we work with an empty shape?

Reply:
When val is scalar and I don't unsqueeze and squeeze I get :

```
val[val >= self.nvec] = (self.nvec.expand_as(val)[val >= self.nvec] - 1).type(
RuntimeError: expand(torch.LongTensor{[1]}, size=[]): the number of sizes provided (0) must be greater or equal to the number of dimensions in the tensor (1)
```